sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -591,18 +591,40 @@ case class MapObjects private(
       case _ => inputData.dataType
     }
 
-    val (getLength, getLoopVar) = inputDataType match {
+    val (getLength, prepareLoop, getLoopVar) = inputDataType match {
       case ObjectType(cls) if classOf[Seq[_]].isAssignableFrom(cls) =>
-        s"${genInputData.value}.size()" -> s"${genInputData.value}.apply($loopIndex)"
+        val it = ctx.freshName("it")
```

Comment:
This would cause some types of `Seq` to be traversed by iterator when they support constant-time random access. That's not so bad. But I think you could make a rule for `IndexedSeq` that takes precedence that retains the old behavior.

Reply:
That's not a big deal, we are traversing the collection from start to end, so either Iterator or accessing by index won't have much difference, may not worth to create 2 branches here.

Reply:
I agreed. Currently the branches are a bit complicated.